Fixed wrong CSync implementation affecting 480i field alignment and c…

e8root opened the pull request:
Fixed wrong CSync implementation affecting 480i field alignment and causing curl on the top of the screen on some CRTs.

kalachx committed:
Fixed wrong CSync implementation affecting 480i field alignment and causing curl on the top of the screen on some CRTs.

marqs85 commented:
It would be best to add an option to select between AND and XNOR type of combiner. I'm afraid using the former could cause issues on some other displays which rely on HSYNC lock during VSYNC. Hopefully these 2 alternatives could cover compatibility for all displays so that equalization pulses (see type C below) would not need to be implemented.

![CSYNC types](https://static1.squarespace.com/static/51f517f0e4b01da70d01ca2a/t/5bd8df8840ec9ab5573140ac/1540939656854/csync_progression.png)

kalachx committed:
Added option to change CSync combiner type to allow selecting legacy XNOR sync combiner. Made space to implement Type B and Type C sync combiners in the future.

e8root commented:
Fair enough. I added option to change the sync combiner type.
For your information the current combiner type is neither of these presented in your diagram. It is actually invalid XNOR method which delays HSync while VSync pulse happens and delays VSync pulse on even fields. When both HS and VS change from 1 to 0 during even field there should be the beginning of both HS and VS pulses. Instead we still get 1 because both signals change at the same time and XORing them won't detect that. HS and VS pulses only begin in this case at the end of HS pulse.

Found good description of XNOR and why it is not valid: [https://www.hdretrovision.com/blog/2019/10/10/engineering-csync-part-2-falling-short](https://www.hdretrovision.com/blog/2019/10/10/engineering-csync-part-2-falling-short)

Implementing proper B and especially C type combiners looks doable but not trivial enough to implement in few minutes.
Just in case I made some space (used two bits) and added AND and XNOR combiners and placeholder for two more CSync combiners. I made "Type A - AND" the default (2'd0) because all of my CRTs work correctly and for the time it seems most proper way to derive csync. It also allows 480i work on HDMI to Component DACs when combined with "HDMI combined sync" option I added in previous PR. I think "Legacy - XNOR" option should be 4th option. Here it will will be at 2'd1 and in the future hopefully seamlessly replaced by proper Type B which should be only more compatible and otherwise used in the same cases where Type A was a no go. I was not sure about user config definition if fw should be at 80 or 81 and what number to use. 110 seems reserved so I chose 111.

BTW. I did not mention it in PR because I did not realize it before but OSSC Pro as is doesn't produce proper 480i with my two HDMI to Component DACs. AND combiner along with HDMI combined sync (note: forgot to commit sc_config_regs.h in my previous PR - sorry!) I get proper 480i on these DACs. Or to be more precise it looks slightly less correct than YPbPr from Legacy AV Out using AND combiner but still better than any of the results using current XNOR method on any Legacy AV Out or HDMI to VGA DACs.

marqs85 commented:
Yeah, a more proper XNOR-esque combiner would need shifting HSYNC backwards in time during VSYNC. It could be tapped "for free" from the beginning of postprocessing pipeline and masked with usual HSYNC to keep resulting falling edge aligned, but this would result to only 11 cycle long inverted HSYNC which I doubt is enough for CRTs.
